### PR TITLE
[logging] Getting instance logs implementation

### PR DESCRIPTION
**Pull request by Eabramov84**
Signed-off-by: Yevgen Abramov <Yevgen_Abramov@epam.com>

### CODE REVIEW

File: logging/logging_test.go [hunk 1]
```diff
@@ -58,118 +52,98 @@ func init() {
...
-	}
+	subject := "subject0"
 
-	if err = startService("logservice0"); err != nil {
-		t.Fatalf("Can't start service: %s", err)
-	}
+	instance := uint64(0)
 
-	time.Sleep(10 * time.Second)
+	instanceFilter := cloudprotocol.InstanceFilter{ServiceID: "logservice0", SubjectID: &subject, Instance: &instance}
```
Comment (al1img) [inline logging/logging_test.go:97]: Create separate function to fill instanceFilter. Or you can create functions to create string and unit pointers and use here in place:
```golang
func newString(value string) *string {
     return &value
}

cloudprotocol.InstanceFilter {SubjectID: newString("subject0")
```
```

File: logging/archivator.go [hunk 2]
```diff
@@ -51,7 +51,9 @@ var errMaxPartCount = errors.New("max part count reached")
  * Private
  ******************************************************************************/
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply logging/archivator.go:52]: done

File: logging/logging.go [hunk 3]
```diff
@@ -15,161 +15,175 @@
...
-// GetServiceLog returns service log
-func (instance *Logging) GetServiceLog(request *pb.ServiceLogRequest) {
-	log.WithFields(log.Fields{
-		"serviceID": request.ServiceId,
-		"logID":     request.LogId,
-		"dateFrom":  request.From,
-		"dateTill":  request.Till}).Debug("Get service log")
+// GetInstanceLog returns instance log.
+func (instance *Logging) GetInstanceLog(request cloudprotocol.RequestServiceLog) error {
+	if data, err := json.Marshal(&request); err == nil {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply logging/logging.go:113]: some fields are nil. A lot of check vars should be added to display valid log.
We do not have a lot of calls to get logs. From this point of view, ut is not so expensive approach.

File: logging/logging_test.go [hunk 4]
```diff
@@ -365,150 +354,285 @@ func TestMaxPartCountLog(t *testing.T) {
...
-	[Service]
-	Type=simple
-	Restart=always
-	RestartSec=1
-	ExecStart=/bin/bash -c 'while true; do echo "[$(date --rfc-3339=ns)] This is log"; sleep 0.1; done'
-	
-	[Install]
-	WantedBy=multi-user.target
-`
+func (provider *testInstanceIDProvider) AddFilter(filter cloudprotocol.InstanceFilter) (instanceID string) {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply logging/logging_test.go:480]: yes. Changed

File: logging/logging.go [hunk 5]
```diff
@@ -15,161 +15,175 @@
...
+// InstanceIDProvider provides instances Id.
+type InstanceIDProvider interface {
+	GetInstanceIDs(cloudprotocol.InstanceFilter) ([]string, error)
 }
 
-// Logging instance
+// Logging instance.
 type Logging struct {
-	logChannel chan *pb.LogData
+	logChannel       chan *cloudprotocol.PushLog
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply logging/logging.go:60]: done

File: logging/logging_test.go (hunk 1 above)
Comment (Eabramov84) [reply logging/logging_test.go:97]: added createInstanceFilter method

File: logging/logging.go [hunk 6]
```diff
@@ -15,197 +15,212 @@
...
-/*******************************************************************************
+/***********************************************************************************************************************
  * Types
- ******************************************************************************/
+ **********************************************************************************************************************/
 
-// ServiceProvider provides service info
-type ServiceProvider interface {
-	GetService(serviceID string) (service launcher.Service, err error)
+// InstanceIDProvider provides instances Id.
```
Comment (al1img) [inline logging/logging.go:53]: ID

File: logging/logging.go [hunk 7]
```diff
@@ -250,69 +276,78 @@ func (instance *Logging) getLog(request getLogRequest) {
...
-			if err == errMaxPartCount {
+			if errors.Is(err, errMaxPartCount) {
 				log.Warn(err)
 				break
 			}
 
-			err = aoserrors.Wrap(err)
-
-			return
+			return err
```
Comment (al1img) [inline logging/logging.go:284]: Here and everywhere wrap with aoserrors all the time.

File: logging/logging.go [hunk 8]
```diff
@@ -250,69 +276,78 @@ func (instance *Logging) getLog(request getLogRequest) {
...
 
 	var crashTime uint64
 
+	crashTime, err = instance.getCrashTime(journal, request.from)
+	if err != nil {
+		return err
+	}
+
+	if crashTime == 0 {
+		return
```
Comment (al1img) [inline logging/logging.go:322]: Empty return. I suggest use anonymous return value in this and other functions.

File: logging/logging_test.go [hunk 9]
```diff
@@ -58,118 +52,93 @@ func init() {
...
 	logging, err := logging.New(&config.Config{Logging: config.Logging{
 		MaxPartSize: 1024, MaxPartCount: 10,
-	}}, &serviceProvider)
+	}}, &instanceProvider)
 	if err != nil {
 		t.Fatalf("Can't create logging: %s", err)
 	}
 	defer logging.Close()
 
 	from := time.Now()
```
Comment (al1img) [inline logging/logging_test.go:90]: Combine all new vars in one var block:
```go
var (
    from = time.Now()
    instanceID = instanceProvider.addFilter(instanceFilter)
    ...
    till = from.Add(5 * time.Second)
)

File: logging/logging_test.go [hunk 10]
```diff
@@ -365,150 +335,276 @@ func TestMaxPartCountLog(t *testing.T) {
...
+	journalEntry.Fields[sdjournal.SD_JOURNAL_FIELD_SYSTEMD_CGROUP] = cgroupUnit
+	journalEntry.Fields[sdjournal.SD_JOURNAL_FIELD_PRIORITY] = priority
+
+	journalEntry.RealtimeTimestamp = uint64(currentTime.UnixNano() / 1000)
+	journalEntry.MonotonicTimestamp = uint64(currentTime.UnixNano() / 1000)
+
+	journal.messages = append(journal.messages, &journalEntry)
+}
+
+func (journal *testSystemdJournal) isMathesEqual(etalonMatches []string) error {
```
Comment (al1img) [inline logging/logging_test.go:558]: Matches?

File: logging/logging_test.go [hunk 11]
```diff
@@ -178,153 +147,154 @@ func TestGetSystemLog(t *testing.T) {
 	from := time.Now()
```
Comment (al1img) [inline logging/logging_test.go:147]: from, till in one var block

File: logging/logging_test.go [hunk 12]
```diff
@@ -178,153 +147,154 @@ func TestGetSystemLog(t *testing.T) {
...
-	if err = createService("logservice2"); err != nil {
-		t.Fatalf("Can't create service: %s", err)
-	}
+	instanceFilter := createInstanceFilter("logservice2", "subject2", 0)
 
-	from := time.Now()
+	_ = instanceProvider.addFilter(instanceFilter)
 
-	time.Sleep(5 * time.Second)
+	from := time.Now()
```
Comment (al1img) [inline logging/logging_test.go:189]: ditto

File: logging/logging_test.go [hunk 13]
```diff
@@ -178,153 +147,154 @@ func TestGetSystemLog(t *testing.T) {
...
+	}, &instanceProvider)
 	if err != nil {
 		t.Fatalf("Can't create logging: %s", err)
 	}
 	defer logging.Close()
 
-	if err = createService("logservice3"); err != nil {
-		t.Fatalf("Can't create service: %s", err)
-	}
+	instanceFilter := createInstanceFilter("logservice3", "subject3", 0)
```
Comment (al1img) [inline logging/logging_test.go:218]: Move new vars into one var block.

File: logging/logging_test.go [hunk 14]
```diff
@@ -178,153 +147,154 @@ func TestGetSystemLog(t *testing.T) {
...
+	logging, err := logging.New(&config.Config{
+		Logging: config.Logging{MaxPartSize: 512, MaxPartCount: 2},
+	}, &instanceProvider)
 	if err != nil {
 		t.Fatalf("Can't create logging: %s", err)
 	}
 	defer logging.Close()
 
-	from := time.Now()
+	instanceFilter := createInstanceFilter("logservice4", "subject4", 0)
```
Comment (al1img) [inline logging/logging_test.go:277]: ditto

File: logging/logging_test.go [hunk 15]
```diff
@@ -365,150 +335,276 @@ func TestMaxPartCountLog(t *testing.T) {
...
 
-func stopService(serviceID string) (err error) {
-	channel := make(chan string)
-
-	if _, err = systemd.StopUnitContext(context.Background(),
-		"aos_"+serviceID+".service", "replace", channel); err != nil {
-		return aoserrors.Wrap(err)
+func (journal *testSystemdJournal) GetEntry() (entry *sdjournal.JournalEntry, err error) {
+	if journal.simulateError {
+		return entry, aoserrors.New("Simulated error")
```
Comment (al1img) [inline logging/logging_test.go:533]: error from small letter.

File: logging/logging_test.go [hunk 16]
```diff
@@ -365,150 +335,276 @@ func TestMaxPartCountLog(t *testing.T) {
...
+	}, &instanceProvider)
+	if err != nil {
+		t.Fatalf("Can't create logging: %s", err)
 	}
+	defer loggingInstance.Close()
 
-	return *s, nil
-}
+	if err := loggingInstance.GetInstanceLog(cloudprotocol.RequestServiceLog{
+		InstanceFilter: cloudprotocol.InstanceFilter{ServiceID: "noService"},
```
Comment (al1img) [inline logging/logging_test.go:354]: Use createInstanceFilter function in all cases for convenience.

File: logging/logging.go [hunk 17]
```diff
@@ -15,197 +15,212 @@
...
+	instances, err := instance.instanceProvider.GetInstanceIDs(request.InstanceFilter)
+	if err != nil {
+		return aoserrors.Wrap(err)
 	}
 
-	if request.From != nil {
-		localTime := request.GetFrom().AsTime()
-		logRequest.from = &localTime
+	if len(instances) == 0 {
+		return aoserrors.New("no instance ids for log request")
```
Comment (al1img) [inline logging/logging.go:123]: ids -> IDs in all the cases.

File: logging/logging_test.go [hunk 18]
```diff
@@ -365,150 +335,276 @@ func TestMaxPartCountLog(t *testing.T) {
...
+		Logging: config.Logging{MaxPartSize: 1024, MaxPartCount: 10},
+	}, &instanceProvider)
+	if err != nil {
+		t.Fatalf("Can't create logging: %s", err)
 	}
+	defer loggingInstance.Close()
 
-	return *s, nil
-}
+	if err := loggingInstance.GetInstanceLog(cloudprotocol.RequestServiceLog{
```
Comment (al1img) [inline logging/logging_test.go:353]: launcher will return empty slice if no service found.

File: logging/logging.go (hunk 8 above)
Comment (Eabramov84) [reply logging/logging.go:322]: accep

File: logging/logging_test.go (hunk 9 above)
Comment (Eabramov84) [reply logging/logging_test.go:90]: done

File: logging/logging_test.go (hunk 11 above)
Comment (Eabramov84) [reply logging/logging_test.go:147]: done

**Review by al1img — APPROVED** [review]

**Pull request merged**